llm4pddl/approaches/llm_approaches.py

```diff
@@ -0,0 +1,158 @@
+    PyperplanType, Task, TaskMetrics
+
+
+class LLMOpenLoopApproach(BaseApproach):
+    """An approach that simply queries the LLM to solve tasks."""
+
+    def __init__(self, num_completions: int, temperature: float) -> None:
+        super().__init__()
+        # Set up the LLM.
+        self._num_completions = num_completions
+        self._temperature = temperature
+        self._llm = OpenAILLM(FLAGS.llm_model_name)
+        # Set after learning.
+        self._prompt_prefix = ""
+
+    @property
+    def is_learning_based(self) -> bool:
+        return True
+
+    @property
+    def is_planning_based(self) -> bool:
+        return False
+
+    def get_name(self) -> str:
+        return "llm-open-loop"
+
+    def solve(self, task: Task) -> Tuple[Optional[Plan], TaskMetrics]:
+        new_prompt = self._create_prompt(task, [])  # empty partial plan
+        prompt = self._prompt_prefix + new_prompt
+        logging.debug(f"Querying with prompt suffix:\n{new_prompt}")
+        responses = self._llm.sample_completions(
+            prompt=prompt,
+            temperature=self._temperature,
+            seed=FLAGS.seed,
+            num_completions=self._num_completions)
+        return self._llm_responses_to_plan(responses, task)
+
+    def train(self, dataset: Dataset) -> None:
+        prompts = []
+        for datum in dataset:
+            prompt = self._create_prompt(datum.task, datum.solution)
+            prompts.append(prompt)
+        self._prompt_prefix = "\n\n".join(prompts) + "\n\n"
+        logging.debug(f"Created prompt prefix:\n{self._prompt_prefix}")
+
+    @staticmethod
+    def _create_prompt(task: Task, plan: Plan) -> str:
+        """Create a prompt entry for a single task and (maybe partial) plan."""
+        # Extract only the objects, init, and goal from the problem file,
+        # stripping out any comments or other extraneous text.
+        domain, problem = utils.parse_task(task)
+        # Create the objects string.
+        type_to_objs: Dict[PyperplanType, List[PyperplanObject]] = {
+            t: []
+            for t in domain.types.values()
+        }
+        for obj in sorted(problem.objects):
+            obj_type = problem.objects[obj]
+            type_to_objs[obj_type].append(obj)
+        objects_strs: List[str] = []
+        for typ, objs in type_to_objs.items():
+            if not objs:
+                continue
+            typ_str = " ".join(objs) + " - " + str(typ)
+            objects_strs.append(typ_str)
+        objects_str = "\n  ".join(objects_strs)
+        # Create the init string.
+        init_str = " ".join(
+            utils.pred_to_str(p) for p in problem.initial_state)
+        # Create the goal string.
+        goal_str = " ".join(utils.pred_to_str(p) for p in problem.goal)
+        # Create the solution string.
+        solution_str = "\n  ".join(plan)
+        prompt = f"""(:objects
+  {objects_str}
+)
+(:init
+  {init_str}
+)
+(:goal
+  {goal_str}
+)
+solution:
```

Comment:
so we are planning to use my cleaning method right before we return the string here, right?

Reply:
right!